tests/unit/db/document_tests.py

```diff
@@ -0,0 +1,501 @@
+
+class DocumentTests(UnitTestDbBase):
+    """
+    CouchDatabase/CloudantDatabase unit tests
+    """
+
+    def setUp(self):
+        """
+        Set up test attributes
+        """
+        super(DocumentTests, self).setUp()
+        self.db_set_up()
+
+    def tearDown(self):
+        """
+        Reset test attributes
+        """
+        self.db_tear_down()
+        super(DocumentTests, self).tearDown()
+
+    def test_constructor_with_docid(self):
+        """
+        Test instantiating a Document providing an id
+        """
+        doc = Document(self.db, 'julia006')
+        self.assertIsInstance(doc, Document)
+        self.assertEqual(doc.r_session, self.db.r_session)
+        self.assertEqual(doc.get('_id'), 'julia006')
+        self.assertEqual(
+            doc.document_url, posixpath.join(self.db.database_url, 'julia006')
+        )
+
+    def test_constructor_without_docid(self):
+        """
+        Test instantiating a Document without providing an id
+        """
+        doc = Document(self.db)
+        self.assertIsInstance(doc, Document)
+        self.assertEqual(doc.r_session, self.db.r_session)
+        self.assertIsNone(doc.get('_id'))
+        self.assertIsNone(doc.document_url)
+
+    def test_document_exists(self):
+        """
+        Test whether a document exists remotely
+        """
+        doc = Document(self.db)
+        self.assertFalse(doc.exists())
+        doc['_id'] = 'julia006'
+        self.assertFalse(doc.exists())
+        doc.create()
+        self.assertTrue(doc.exists())
+
+    def test_retrieve_document_json(self):
+        """
+        Test the document dictionary renders as json appropriately
+        """
+        doc = Document(self.db)
+        doc['_id'] = 'julia006'
+        doc['name'] = 'julia'
+        doc['age'] = 6
+        doc_as_json = doc.json()
+        self.assertIsInstance(doc_as_json, str)
+        self.assertEqual(json.loads(doc_as_json), doc)
+
+    def test_create_document_with_docid(self):
+        """
+        Test creating a document providing an id
+        """
+        doc = Document(self.db, 'julia006')
+        doc['name'] = 'julia'
+        doc['age'] = 6
+        self.assertFalse(doc.exists())
+        self.assertIsNone(doc.get('_rev'))
+        doc.create()
+        self.assertTrue(doc.exists())
+        self.assertTrue(doc.get('_rev').startswith('1-'))
+
+    def test_create_document_without_docid(self):
+        """
+        Test creating a document remotely without providing an id
+        """
+        doc = Document(self.db)
+        doc['name'] = 'julia'
+        doc['age'] = 6
+        self.assertFalse(doc.exists())
+        self.assertIsNone(doc.get('_id'))
+        self.assertIsNone(doc.get('_rev'))
+        doc.create()
+        self.assertTrue(doc.exists())
+        self.assertIsNotNone(doc.get('_id'))
+        self.assertTrue(doc.get('_rev').startswith('1-'))
+
+    def test_create_document_with_docid(self):
```

Comment:
Duplicate method name. I'm surprised this doesn't throw some error. Maybe it's valid for unittest? Either way it's a bit confusing.